source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -1,11 +1,129 @@
 #include "extensions/filters/http/ext_proc/ext_proc.h"
 
+#include "extensions/filters/http/ext_proc/mutation_utils.h"
+
+#include "absl/strings/str_format.h"
+
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace ExternalProcessing {
 
-void Filter::onDestroy() {}
+using envoy::service::ext_proc::v3alpha::ProcessingRequest;
+using envoy::service::ext_proc::v3alpha::ProcessingResponse;
+
+using Http::FilterHeadersStatus;
+using Http::RequestHeaderMap;
+
+static const std::string kErrorPrefix = "ext_proc error";
+
+void Filter::closeStream() {
+  if (!stream_closed_) {
+    if (stream_) {
+      ENVOY_LOG(debug, "Closing gRPC stream to processing server");
+      stream_->close();
+      stats_.streams_closed_.inc();
+    }
+    stream_closed_ = true;
+  }
+}
+
+void Filter::onDestroy() { closeStream(); }
+
+FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool end_of_stream) {
+  // We're at the start, so start the stream and send a headers message
+  request_headers_ = &headers;
+  stream_ = client_->start(*this, config_->grpcTimeout());
+  stats_.streams_started_.inc();
+  ProcessingRequest req;
+  auto* headers_req = req.mutable_request_headers();
+  MutationUtils::buildHttpHeaders(headers, *headers_req->mutable_headers());
+  headers_req->set_end_of_stream(end_of_stream);
+  request_state_ = FilterState::HEADERS;
+  stream_->send(std::move(req), false);
```

Comment:
What about `attributes` as specified in the configuration by `request_attributes` and `response_attributes`? Also out of scope for this PR?

Reply:
Thanks for looking at it and I'm glad that you're excited. The filter is currently marked "work in progress" and all the configuration is currently marked "not implemented yet" because it's likely to take a few PRs to get a reasonable part of the implementation working. Everything that you're asking about is in scope for this filter, but I'm trying to implement it as a series of smaller PRs rather than a huge one that implements all the parts. I'm expecting that it'll take several more weeks until this has enough of the surface area implemented to be actually useful.